.github/workflows/refapp-3x-login.yml

```
@@ -1,17 +1,43 @@
 name: RefApp 3.x Login
 on:
+  push:
+    branches: [master]
   pull_request:
-    branches: [ master ]
+    branches: [master]
+  repository_dispatch:
+    types: [qa]
+  workflow_dispatch:
 jobs:
   build:
     runs-on: ubuntu-latest
+    services:
+      firefox:
+        image: selenium/node-firefox:3.141.59-gold
+        env:
+          HUB_HOST: hub
+          HUB_PORT: 4444
     steps:
-      - uses: actions/checkout@v2
+      - name: Set up JDK 1.8
+        uses: actions/setup-java@v1
+        with:
+          java-version: 1.8
+      - name: Checkout qaframework
+        uses: actions/checkout@master
+        with:
+          repository: ${{github.repository}}
+      - name: run db and web containers
+        run: |
+          docker-compose -f docker/docker-compose-refqa.yml up -d
+      - name: wait for openmrs instance to start
+        run: while [[ "$(curl -s -o /dev/null -w ''%{http_code}'' http://localhost:8080/openmrs/login.htm)" != "200" ]]; do sleep 1; done
       - name: Using Node.js
         uses: actions/setup-node@v1
         with:
           node-version: '12.x'
       - name: Installing dependencies
         run: npm install
-      - name: Run patient registration workflow tests
+      - name: Run RefApp 3.x frontend
+        run: npx openmrs@next start --backend localhost:8080 --port 8081 &
```

Per discussion on Slack,
```suggestion
        run: npx openmrs@next start --backend http://localhost:8080 --port 8081 &
```

---

Thanks @brandones ! 
I have done the necessary changes. It works fine in my local setup. But the following issue occurs in the Github workflow:
`Error: connect ECONNREFUSED 127.0.0.1:8081`


also, I tried running the test and the proxy in a single step as well. But the error is the same.
```
- name: Run patient login workflow tests
        run: |
                npx openmrs@next start --backend http://localhost:8080 --port 8081 &
                npm run refApp3Login
```

---

I think [this SO Q/A](https://stackoverflow.com/a/62546833/1464495) is about this same issue. It seems like we need to publish the image and use it as a service container: https://docs.github.com/en/actions/guides/about-service-containers

---

Thanks @brandones,

Necessary changes were done to the workflow file:
https://github.com/openmrs/openmrs-contrib-qaframework/pull/50/commits/d522a0f602965f5c44af6a3db3f1f5a1c4dbe31b

Had to remove some parameters such as `command` `healthcheck` and `depends_on` because Github doesn't support them. 

Now it throws the following error:
```
Exit code 1 returned from process: file name '/usr/bin/docker', arguments 'create --name ce9c6e7005dd4f9bba8f5188bffc9c5d_openmrsopenmrsreferenceapplicationdistrodemo_516d6f --label 3bec5f --network github_network_4f55a29bcb3942ffa425949bc63bcea0 --network-alias openmrs-referenceapplication -p 8080:8080  -e "DB_DATABASE=openmrs" -e "DB_HOST=openmrs-referenceapplication-mysql" -e "DB_USERNAME=openmrs" -e "DB_PASSWORD=Admin123" -e "DB_CREATE_TABLES=true" -e "DB_AUTO_UPDATE=false" -e "MODULE_WEB_ADMIN=false" -e GITHUB_ACTIONS=true -e CI=true -v "openmrs-referenceapplication-data":"/usr/local/tomcat/.OpenMRS" -v "./activemq-data":"/usr/local/tomcat/.OpenMRS/activemq-data" openmrs/openmrs-reference-application-distro:demo'.
```

https://github.com/openmrs/openmrs-contrib-qaframework/runs/3184870317?check_suite_focus=true

---

Btw, I'm curious how 2.x tests access the localhost. 

However, I think the container runs fine because `Wait for Openmrs instance to start` stage in the workflow works fine.

Can't it be an issue with the frontend proxy? 🤔

---

Looks like the actual error is `Error response from daemon: create ./activemq-data: "./activemq-data" includes invalid characters for a local volume name, only "[a-zA-Z0-9][a-zA-Z0-9_.-]" are allowed. If you intended to pass a host directory, use absolute path`. So `./activemq-data` is not okay because it's a relative path. Try making it an absolute path.

---

Yeah I have the same curiosity about the 2.x tests. They definitely do work. Not sure why Selenium is able to access the container but we're not able to set up a proxy to it.

Anyway let's keep on with the `services` approach, this seems promising.

---

Thanks @brandones , I have fixed the path. The container is now running. However, it throws the `Error: connect ECONNREFUSED 127.0.0.1:8081` error again. What could go wrong?

---

I don't know, I'm not an expert in GitHub Actions. I need you to research this problem using Google (or another search engine) and on Stack Overflow and try things until you are able to solve it. That is what I would do. I don't think that we are trying to do something impossible. But if you decide that we need to take a different approach, please lead us in that direction.